From 1cacb78f3fef37c6c4bf746c6756f5ae57e46149 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Mon, 12 Aug 2024 21:46:08 +0530
Subject: [PATCH 3/6] wifi: ath12k: optimise skb callback structure
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Although the ‘bool’ type typically consumes 8 bits, the original
design was meant to use only 1 bit. By optimizing the ‘ath12k_skb_rxcb’
with bit fields reduce memory usage from 32 bytes to 24 bytes.
This efficiency gain frees up space for additional members that are
then used to enhance the ultra-fast receive path.

Patch-Dependency: none

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h | 25 ++++++++++++-------------
 1 file changed, 12 insertions(+), 13 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 45fa3c0..99a2323 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -146,20 +146,19 @@ struct ath12k_skb_cb {
 
 struct ath12k_skb_rxcb {
 	dma_addr_t paddr;
-	bool is_first_msdu;
-	bool is_last_msdu;
-	bool is_continuation;
-	bool is_mcbc;
-	bool is_eapol;
-	bool is_end_of_ppdu;
 	struct hal_rx_desc *rx_desc;
-	u8 err_rel_src;
-	u8 err_code;
-	u8 hw_link_id;
-	u8 unmapped;
-	u8 is_frag;
-	u8 tid;
-	u16 peer_id;
+	u8 err_rel_src:3,
+	   err_code:5;
+	u8 hw_link_id:4,
+	   is_first_msdu:1,
+	   is_last_msdu:1,
+	   is_continuation:1,
+	   is_mcbc:1;
+	u8 is_eapol:1,
+	   is_end_of_ppdu:1,
+	   is_frag:1,
+	   tid:4;
+	u16 peer_id:14;
 };
 
 enum ath12k_hw_rev {
-- 
2.34.1

